File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/python/08.Count set bit from numbers 1 to n.py
"""QUESTION:
Given N, return the total number of set bits in all integers from 1 to N (inclusive).

Approach:
Let x = floor(log2(n)). Then use:
countSetBits(n) = (2^{x-1} * x) + (n - 2^x + 1) + countSetBits(n - 2^x).

Time Complexity: O(log n)
Space Complexity: O(log n)
"""
import math

def count_set_bits(n: int) -> int:
    if n <= 1:
        return n
    x = int(math.log2(n))
    highest_power = 1 << x
    upto_power_contribution = (highest_power >> 1) * x
    msb_contribution = n - highest_power + 1
    remainder = n - highest_power
    return upto_power_contribution + msb_contribution + count_set_bits(remainder)

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/javascript/08.Count set bit from numbers 1 to n.js
/**
QUESTION:
Given `N`, compute the total number of set bits in all integers from 1 to N (inclusive).

Approach:
Use the recursive formula based on the highest power of two <= N:
countSetBits(n) = (2^{x-1} * x) + (n - 2^x + 1) + countSetBits(n - 2^x),
where x = floor(log2(n)).

Time Complexity: O(log n)
Space Complexity: O(log n) recursion depth
*/

function countSetBits(n) {
    if (n <= 1) return n;
    const x = Math.floor(Math.log2(n));
    const highestPower = 1 << x;
    const uptoPowerContribution = (highestPower >> 1) * x;
    const msbContribution = n - highestPower + 1;
    const remainder = n - highestPower;
    return uptoPowerContribution + msbContribution + countSetBits(remainder);
}

module.exports = { countSetBits };
